File: dev/requirements/logic/cases/tap-told-card.5.5.case.swift
import Foundation
import SetlistCore

/// 5.5 — Tapping a told card re-opens it as live and opens a second segment for
/// it, leaving its first segment intact.
let tapToldCard_5_5 = RequirementCase(id: "5.5") {
    let van = Joke(title: "Dad's van", body: "…")
    let airport = Joke(title: "Airport security", body: "…")
    var stage = StageModel(jokes: [van, airport], plannedLength: 1200)

    stage.tapCard(jokeID: van.id, at: CaptureTime(0))
    stage.tapCard(jokeID: airport.id, at: CaptureTime(125))
    let firstVanSegment = stage.segments[0]

    // The callback lands and the comedian goes back to the van.
    let outcome = stage.tapCard(jokeID: van.id, at: CaptureTime(209))

    try expectEqual(outcome, .switched(to: van.id), "tap outcome")
    try expectEqual(stage.liveJokeID, van.id, "the re-opened joke is live")
    try expectEqual(stage.segments.count, 3, "the return opened a third segment")
    try expectEqual(stage.segments[0], firstVanSegment, "the first segment is untouched")
    try expectEqual(stage.segments[2].jokeID, van.id, "the third segment belongs to the re-opened joke")
    try expectEqual(stage.segments[2].start, CaptureTime(209), "the third segment starts at the tap")
}

File: dev/requirements/logic/cases/tap-told-card.5.5.case.dart
import 'package:setlist_core/setlist_core.dart';

import '../../shared/expectations.dart';
import '../../shared/requirement_case.dart';

/// 5.5 — Tapping a told card re-opens it as live and opens a second segment for it,
/// leaving its first segment intact.
final tapToldCard_5_5 = RequirementCase('5.5', (tester) {
  final stage = StageModel(
    jokes: [
      Joke(id: 'opener', title: 'Opener', body: ''),
      Joke(id: 'van', title: "Dad's van", body: ''),
    ],
    plannedLength: const Duration(minutes: 20),
  );

  stage.tapCard(jokeId: 'van', at: const CaptureTime(0));
  stage.tapCard(jokeId: 'opener', at: const CaptureTime(60));

  // The callback lands, so the comedian goes back to the van.
  stage.tapCard(jokeId: 'van', at: const CaptureTime(200));

  final vanSegments = [for (final s in stage.segments) if (s.jokeId == 'van') s];
  expectThat(vanSegments.length, 2, 'the van has a second segment');
  expectThat(vanSegments.first.start, const CaptureTime(0), 'its first starts where it did');
  expectThat(vanSegments.first.end, const CaptureTime(60), 'and ends where it did');
  expectAbsent(vanSegments.last.end, 'while the new one is open');
  expectThat(stage.liveJokeId, 'van', 'and the van is live again');
});
